### DISCUSSION TITLE
Reading request body in middleware without consuming It?

### DISCUSSION THREAD

**Discussion by Manouchehri** [honojs/hono#1417] [Q&A]
# Question

> How can I read the request body in middleware without consuming it, so it's still accessible for subsequent handlers?

Question taken from: https://discord.com/channels/1011308539819597844/1011308539819597847/1148952087070265416

# Answer

```typescript
// This is an example middleware that consumes the request body and then creates a new request with the same body
app.post('*', async (c, next) => {
  const incoming_request = c.req.raw;

  const incoming_ab = await incoming_request.arrayBuffer();

  // TODO: Add your real signature verification on incoming_ab here
  const signature_check_passed = true;

  if(!signature_check_passed) {
    return c.json({
      success: false,
      error: "Invalid signature"
    }, 401)
  }

  const outgoing_request = new Request(incoming_request.url, {
    method: "POST",
    headers: incoming_request.headers, // If your signature verification checks the headers, you should provide a sanitized version here
    body: incoming_ab
  })

  c.req.raw = outgoing_request

  await next()
})

app.post('/demo', async (c) => {
  console.debug(await c.req.text());
  return c.text('okay')
})
```

**Comment by ambergristle** [comment]
from what i can tell, hono parses the body via array buffer, updating a cache after consuming the body; see https://github.com/honojs/hono/issues/1499. using the hono validator helper code as reference, i arrived at the following solution, which populates the cached buffer in the middleware, allowing the body to also be consumed in a handler

https://github.com/honojs/hono/blob/81264a9cf5837462c9af6b5653ea1675759fa8db/src/validator/validator.ts#L81

initially i tried to simply clone the request, but the body would get stripped. this may be related to bun: https://github.com/oven-sh/bun/issues/6348. in any case, the below should work

```javascript
app.use(async (c, next) => {
  const arrayBuffer = await c.req.arrayBuffer();
  c.req.bodyCache.arrayBuffer = arrayBuffer;
  
  const body = await new Response(arrayBuffer).text()
  
  await next()
})

app.get('/', (c) => {
  const body = await c.req.json()
  console.log(body) // { ... }
  
  return c.text('ok')
})
```
